File: shared/selfplay_config.py
"""Shared Python-side config dataclasses for AlphaZero-style self-play.

These mirror the Rust structs in `core_game::selfplay_config` and reduce
the long flat-kwargs explosion in each game's `SelfPlayTrainer.run(...)`
signature and CLI → orchestrator → pyo3-session forwarding chain.

Each dataclass has a `session_kwargs()` method that returns the dict
to splat into the corresponding `*SelfPlaySession(...)` pyo3 constructor,
so the boundary between Python config and Rust pyo3 kwargs is one place
to maintain.
"""

from __future__ import annotations

from dataclasses import dataclass, asdict


@dataclass
class MctsConfig:
    """MCTS-related knobs that every game's self-play loop uses identically."""
    simulations: int = 200
    c_puct: float = 1.5
    dir_alpha: float = 0.3
    dir_epsilon: float = 0.25
    forced_playouts: bool = False
    draw_contempt: float = 0.0
    play_batch_size: int = 8
    temperature: float = 1.0
    temp_threshold: int = 20

    def session_kwargs(self) -> dict:
        # Pyo3 session kwargs match field names 1:1.
        return asdict(self)


@dataclass
class PlayoutCapConfig:
    """KataGo-style playout-cap randomization. p == 0.0 disables it."""
    p: float = 0.0
    fast_cap: int = 30

    def session_kwargs(self) -> dict:
        # Pyo3 session prefixes p as `playout_cap_p`.
        return {"playout_cap_p": self.p, "fast_cap": self.fast_cap}


@dataclass
class OpeningRandomConfig:
    """Random-opening-moves window. Each game samples uniformly in [min, max]."""
    min: int = 0
    max: int = 0

    @classmethod
    def from_cli_arg(cls, arg: int | tuple[int, int]) -> "OpeningRandomConfig":
        """Build from `--random-opening-moves N` (int) or `N-M` (tuple)."""
        if isinstance(arg, tuple):
            return cls(min=arg[0], max=arg[1])
        return cls(min=arg, max=arg)

    def session_kwargs(self) -> dict:
        return {
            "random_opening_moves_min": self.min,
            "random_opening_moves_max": self.max,
        }

File: rust/crates/core-game/src/selfplay_config.rs
//! Shared configuration structs for AlphaZero-style self-play across games.
//!
//! These types cover the knobs that every game's `play_selfplay_core` consumes
//! identically — MCTS settings, playout-cap randomization, and opening-move
//! randomization. Game-specific extras (resignation, calibration, max-moves
//! caps, heuristic eval, …) stay in each game's own session config struct.
//!
//! At the pyo3 boundary the Python signature stays flat (individual kwargs),
//! and the binding layer unpacks them into these structs.

/// MCTS-related knobs that every game's self-play loop uses identically.
#[derive(Clone, Copy, Debug)]
pub struct MctsConfig {
    pub simulations: usize,
    pub c_puct: f32,
    pub dir_alpha: f32,
    pub dir_epsilon: f32,
    pub forced_playouts: bool,
    pub draw_contempt: f32,
    /// When true, each self-play game randomly designates one side as the
    /// "contempt side"; `draw_contempt` only applies on that side's turns,
    /// while the other side searches with contempt 0. When false, contempt
    /// is applied symmetrically on every turn (the historical default).
    pub asymmetric_contempt: bool,
    /// Number of leaves collected across games per inference batch.
    pub play_batch_size: usize,
    /// Visit-count temperature applied for the first `temp_threshold` plies.
    pub temperature: f32,
    pub temp_threshold: u32,
}

/// KataGo-style playout-cap randomization. `p == 0.0` disables it; otherwise
/// each turn flips a coin: with probability `p`, run the full `simulations`
/// budget (full-search turn → trains policy + value); else run `fast_cap`
/// sims (value-only training).
#[derive(Clone, Copy, Debug, Default)]
pub struct PlayoutCapConfig {
    pub p: f32,
    pub fast_cap: usize,
}

impl PlayoutCapConfig {
    pub fn enabled(&self) -> bool {
        self.p > 0.0
    }
}

/// Opening-move randomization: each game plays a uniform random count
/// in `[min, max]` of random moves before MCTS takes over. `max == 0`
/// disables it.
#[derive(Clone, Copy, Debug, Default)]
pub struct OpeningRandomConfig {
    pub min: u32,
    pub max: u32,
}

impl OpeningRandomConfig {
    pub fn enabled(&self) -> bool {
        self.max > 0
    }
}
